### PR TITLE
[enhancement] IPFS pinning

### PR DESCRIPTION
**Pull request by sha14r**
## Description

https://www.notion.so/hicommonwealth/IPFS-Pinning-Solution-for-CW-a4c55804a37e47408a72d61f2575ff8d

## How has this been tested?

I hit the route with some sample data and that got pinned to pinata: 
image

Pinning: 
image

DB Entry:

image

Migration undo:
image

image

Migration: 
image
image

## Have proper tags been added (for bug, enhancement, breaking change)?
- [ ] yes

## Does this PR affect any server routes?
- [ ] yes, and they are tested: [enter the % coverage here]
- [ ] yes, and they are not tested: [enter the % coverage here]
- [ ] yes, but I did not run tests
- [ ] no

### CODE REVIEW

File: server/routes/pinIPFS.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+import { DB, sequelize } from '../database';
+import ipfs from '../util/ipfs';
+
+const pinIPFS = async (
+  models: DB,
+  req: Request,
+  res: Response,
+  next: NextFunction
+) => {
+  try {
```
Comment (zakhap) [inline server/routes/pinIPFS.ts:11]: We need to validate the inputs here! 
Make sure the user is logged in (req.user), make sure we're unpacking whatever is being passed into the `req.body`, if we are verifying that it's a jsonblob or a string or... or if we're not validating, just make sure it's not empty.

File: server/models/ipfspins.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+export type IpfsPinsModelStatic = ModelStatic<IpfsPinsInstance>;
+
+export default (
+  sequelize: Sequelize.Sequelize,
+  dataTypes: typeof DataTypes,
+): IpfsPinsModelStatic => {
+  const IpfsPins = <IpfsPinsModelStatic>sequelize.define('IpfsPins', {
+    id: { type: dataTypes.INTEGER, primaryKey: true },
+    ipfs_hash: { type: dataTypes.STRING, allowNull: false },
+  }, {
```
Comment (zakhap) [inline server/models/ipfspins.ts:23]: we should also store the `address_id` of the user's active address when submitting the data to be pinned to IPFS

File: server/routes/pinIPFS.ts (hunk 1 above)
Comment (zakhap) [reply server/routes/pinIPFS.ts:11]: also validate that the `req.body.address_id` is not empty.

File: server/routes/pinIPFS.ts [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+    const ipfsHash = await ipfs(req.user.id, req.body);
```
Comment (zakhap) [inline server/routes/pinIPFS.ts:12]: we should not just push the entirety of the req.body, we should make sure the client submits it inside a variable like `req.body.blob`, we need to make sure `blob` is not null, and then submit that `req.body.blob`.

**Review by zakhap — CHANGES_REQUESTED** [review]
left some notes!

File: server/models/ipfspins.ts (hunk 2 above)
Comment (sha14r) [reply server/models/ipfspins.ts:23]: image
Fixed this

File: server/routes/pinIPFS.ts (hunk 3 above)
Comment (sha14r) [reply server/routes/pinIPFS.ts]: Fixed this, thanks

File: server/routes/pinIPFS.ts (hunk 1 above)
Comment (sha14r) [reply server/routes/pinIPFS.ts:19]: Inputs are validated

File: server/models/ipfspins.ts (hunk 2 above)
Comment (zakhap) [reply server/models/ipfspins.ts]: it's not in the model defined here

File: server/migrations/20220228155010-add-ipfs-pins.js [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+      id: {
+        allowNull: false,
+        autoIncrement: true,
+        primaryKey: true,
+        type: Sequelize.INTEGER
+      },
+      ipfs_hash: {
+        type: Sequelize.STRING
+      },
+      created_at: { type: Sequelize.DATE, allowNull: false },
```
Comment (zakhap) [inline server/migrations/20220228155010-add-ipfs-pins.js:20]: add address_id to model migration

**Review by zakhap — CHANGES_REQUESTED** [review]

File: server/migrations/20220228155010-add-ipfs-pins.js (hunk 4 above)
Comment (sha14r) [reply server/migrations/20220228155010-add-ipfs-pins.js:20]: It is in line number 12

File: server/models/ipfspins.ts (hunk 2 above)
Comment (sha14r) [reply server/models/ipfspins.ts]: I'm not sure why you don't see it. Here it is in line number 7 &25
image

**Pull request closed without merging**